### ISSUE TITLE
Feature: Burn equivalent to torch.retain_grad

### ISSUE THREAD

**Issue by ArthurBrussee** [tracel-ai/burn#1802]
### Feature description

I'm writing some tests to check gradients again a reference implementation. This works great for leaf nodes, but I can't atm seem to get gradients of intermediate nodes. PyTorch solves this with [my_tensor.retain_grad()](https://pytorch.org/docs/stable/generated/torch.Tensor.retain_grad.html), which instructs the autodiff engine to keep the gradients during the backward pass. An equivalent in Burn could help with this.

### Feature motivation

Testing of gradient activations.

### Suggest a Solution

An exact equivalent a la `my_tensor.retain_grad()`, or, alternatively, make `my_tensor.require_grad()` valid on non-leaf nodes (currently panics). The semantics of retain/require are sligthly different, but, the use-cases for retained-but-only-if-calculated gradients don't seem that massive to me... not sure!

**Comment by Apogeum12** [comment]
Any updates with these features? Maybe it's available now in burn, so then how's it possible to use it? I'm working on Gan, and it would be nice if I could use this feature, like in pytorch. Because currently I have to run the generator twice in one iteration to generate fake data in one iteration, in other cases, when I clone fake data, the generator loss grows into NaN after a couple iterations.

**Comment by AsherJingkongChen** [comment]
> Any updates with these features? Maybe it's available now in burn, so then how's it possible to use it? I'm working on Gan, and it would be nice if I could use this feature, like in pytorch. Because currently I have to run the generator twice in one iteration to generate fake data in one iteration, in other cases, when I clone fake data, the generator loss grows into NaN after a couple iterations.

I am implementing my own backward propagation, and I have found a similar approach to Torch's `tensor.retain_grad()`. However, it is a bit complicated.

First, I can implement `Backward` trait for my own operator, where I need to register my derived gradients:

```rust
use burn::backend::autodiff::{
    checkpoint::{base::Checkpointer, strategy::NoCheckpointing},
    grads::Gradients,
    ops::{Backward, Ops, OpsKind},
    NodeID,
};

#[derive(Debug)]
struct MY_BACKWARD;

// 2-D Current node, 1 Parent node
impl<B: Backend> Backward<B, 2, 1> for MY_BACKWARD
{
    type State = (NodeID, NodeID);
    
    fn backward(
        self,
        ops: Ops<Self::State, 1>,
        grads: &mut Gradients,
        _checkpointer: &mut Checkpointer,
    ) {
        // TODO: Computing gradients of parents

        // Add entry (ops.state.YOUR_NODE_ID, YOUR_TENSOR_GRAD) to grads' internal container
        grads.register::<B, 2>(ops.state.YOUR_NODE_ID, YOUR_TENSOR_GRAD);
    }
}
```

Second, remove out the gradient tensors after calling `Tensor::backward`:

```rust
// Deriving the output and get a non-leaf tensor
let (output, YOUR_TENSOR) = model.forward(YOUR_PARAM);

// Computing gradients
let loss = YOUR_LOSS_FUNCTION(output, reference);
let mut grads = loss.backward();

// Extract the gradient of tensor
let YOUR_TENSOR_GRAD = YOUR_TENSOR.grad_remove(&mut grads);
```

To implement `model.forward`, you can check the `burn-autodiff` source code: https://github.com/tracel-ai/burn/blob/756931211c9e1343fe21363fa0febce45869ffb9/crates/burn-autodiff/src/ops/module.rs#L286

Also, it is recommended to see the chapter 11.1.1 of the book here: https://burn.dev/burn-book/advanced/backend-extension/custom-wgpu-kernel.html

**Comment by Apogeum12** [comment]
@AsherJingkongChen Wow, very good implementation. Maybe when I'll more time I investigate it. thanks
